chore: Add sentry to both client and server

StefanNemeth opened the pull request:
Adding Sentry to our client and server to gain real-time visibility into errors and performance issues, ensuring quicker detection and resolution of problems. This will improve overall system reliability, user experience, and streamline debugging processes for both client and server components.

I also added a Github Action file that uploads the current source map of the main branch to Sentry so debugging is easier.

egekocabas commented on .github/workflows/sentry.yml:
```diff
@@ -0,0 +1,41 @@
+          sourcemaps: './client/dist/client'
+        env:
+          SENTRY_AUTH_TOKEN: ${{ secrets.SENTRY_AUTH_TOKEN }}
+          SENTRY_ORG: ${{ vars.SENTRY_ORG }}
+          SENTRY_PROJECT: ${{ vars.SENTRY_ORG }}
```
Is it an accident or are we using the same variable `SENTRY_ORG` for both **org** and **project**? because i also see a variable called `SENTRY_PROJECT` in the repository environments.

StefanNemeth replied:
Good catch!

egekocabas commented on .github/workflows/sentry.yml:
I have never used sentry so i actually don't know what is going on but it looks like we are sending the generated client code to sentry but do we also need to send the generated jar?

StefanNemeth replied:
When building the client code, the code is optimized and minified so the users receive a small package. Sending the source map (the mapping between the optimized code and the source code) to Sentry gives us better debugging: When there is a JavaScript error, we immediately see the code that caused it

egekocabas commented on client/package.json:
```diff
@@ -6,7 +6,8 @@
     "lint": "eslint --max-warnings 0 --fix .",
-    "generate:openapi": "openapi-ts"
+    "generate:openapi": "openapi-ts",
+    "sentry:sourcemaps": "sentry-cli sourcemaps inject --org helios-jf --project helios-client ./dist/client && sentry-cli sourcemaps upload --org helios-jf --project helios-client ./dist/client"
```
instead of hardcoded **org** and **project** can we use the environment variables?

StefanNemeth replied:
Removed it, see my comment

egekocabas commented on client/src/main.ts:
```diff
@@ -1,5 +1,21 @@
+
+if (environment.production) {
+  Sentry.init({
+    dsn: 'https://0ef8bdd2438d4ab0333984c1c3601829@o4508675143499777.ingest.de.sentry.io/4508675194290256',
```
Exposing this DSN directly in a public repository doesn't seem safe. Could this allow unauthorized access or misuse? Maybe storing it in an environment variable would be a more secure approach?

StefanNemeth replied:
What do you mean? It's not a secret and will be public by default. The client/browser will have to know where to send the request to.

egekocabas replied:
I wasn’t sure how Sentry works, so I was just asking 😅  If the DSN needs to be in the client code, that makes sense since everything on the client side is public anyway. Thanks for clarifying.

StefanNemeth replied:
Still moved it over to the environment config file so it's cleaner :)

egekocabas commented on client/src/main.ts:
```diff
@@ -1,5 +1,21 @@
+if (environment.production) {
+  Sentry.init({
+    dsn: 'https://0ef8bdd2438d4ab0333984c1c3601829@o4508675143499777.ingest.de.sentry.io/4508675194290256',
+    integrations: [Sentry.browserTracingIntegration(), Sentry.replayIntegration()],
+    // Tracing
+    tracesSampleRate: 1.0, //  Capture 100% of the transactions
+    // Set 'tracePropagationTargets' to control for which URLs distributed tracing should be enabled
+    tracePropagationTargets: ['localhost'],
+    // Session Replay
+    replaysSessionSampleRate: 0.1, // This sets the sample rate at 10%. You may want to change it to 100% while in development and then sample at a lower rate in production.
+    replaysOnErrorSampleRate: 1.0, // If you're not already sampling the entire session, change the sample rate to 100% when sampling sessions where errors occur.
```
Can you rewrite the comments in here so that we know what each property does?

egekocabas commented on server/application-server/build.gradle:
```diff
@@ -134,6 +135,17 @@ openApi {
     }
 }
 
+sentry {
+  // Generates a JVM (Java, Kotlin, etc.) source bundle and uploads your source code to Sentry.
+  // This enables source context, allowing you to see your source
+  // code as part of your stack traces in Sentry.
+  includeSourceContext = true
+
+  org = "helios-jf"
+  projectName = "helios-server"
+  authToken = System.getenv("SENTRY_AUTH_TOKEN")
+}
+
```
What does this configuration do exactly?
Does it require a specific command to run, like `gradlew sentry`?
Does it execute during the build step or at runtime?
If it's part of the build step, do we need to set the SENTRY_AUTH_TOKEN environment variable in the production build workflow as well?

StefanNemeth replied:
We will need to add SENTRY_AUTH_TOKEN to the environment variables of the production server. It's not supposed to run locally.

egekocabas replied:
Can you add the token in prod environment secrets and also update this [deploy_docker.yml](https://github.com/ls1intum/Helios/blob/main/.github/workflows/deploy_docker.yml#L68-L95) section?

egekocabas replied:
Would be also good to mention this token in the [documentation](https://github.com/ls1intum/Helios/tree/main/docs)

StefanNemeth replied:
Removed this to simplify things. Now only using the dependency instead of the plugin.

egekocabas commented on server/application-server/src/main/resources/application-prod.yml:
```diff
@@ -44,3 +44,7 @@ monitoring:
     repository-sync-cron: "0 0 * * * *"
+
+sentry:
+  dsn: https://a07d7c305924ed8246351dc5f11864a0@o4508675143499777.ingest.de.sentry.io/4508675150118992
```
Same concern with the client

> Exposing this DSN directly in a public repository doesn't seem safe. Could this allow unauthorized access or misuse? Maybe storing it in an environment variable would be a more secure approach?

StefanNemeth replied:
In this case we could potentially move it over to an environment variable, but shouldn't be a safety risk either.

StefanNemeth replied:
Updated it to use an environment variable

egekocabas commented on server/application-server/src/main/resources/application-prod.yml:
```diff
@@ -44,3 +44,7 @@ monitoring:
+
+sentry:
+  dsn: https://a07d7c305924ed8246351dc5f11864a0@o4508675143499777.ingest.de.sentry.io/4508675150118992
+  traces-sample-rate: 1.0
```
Can you also add a comment about what is this property for?

StefanNemeth replied:
Basically enabled tracing, but removed it so only errors are reported.

egekocabas reviewed (changes requested):
Thanks for integrating Sentry 👍🏻 

I added some comments and concerns about the changes. 

Also, for Sentry, could you invite the team so we can all access it and track issues?

bensofficial commented on client/src/main.ts:
```diff
@@ -1,5 +1,21 @@
+
+if (environment.production) {
+  Sentry.init({
+    dsn: 'https://0ef8bdd2438d4ab0333984c1c3601829@o4508675143499777.ingest.de.sentry.io/4508675194290256',
```
We got our own Sentry instance on https://sentry.ase.in.tum.de. Is there any reason why you don't use that instead of using the cloud sentry?

StefanNemeth replied:
Will do! Thanks. I wasn't aware of our instance.

StefanNemeth commented:
- I simplified the setup without the source maps as they are not really required for now. Replaced the Sentry plugin with the dependency for the server.
- Added comments where it made sense
- Used an environment variable for the server DSN
- Setup two Sentry projects on our own instance and configured the client to use it

egekocabas reviewed (approved):
Thanks for the PR 👍🏻  It looks good to me, couple of thoughts

- I didn't know how to test it so I just run it locally.
- Also couldn't find the project in sentry, i think it should be hidden or secret so thats fine
